### PR TITLE
Added Timetamp

### PR DESCRIPTION
**Pull request by leohuang2013**
Added timestamp for greed search and modified beam search decoder. 

SherpaNcnnResult now includes tokens, timestamps and count besides of existing text field.  Tokens and timestamps are stored in continuous memory in which each token and timestamp ends with nullptr. Use count and pointer increment to get each token and timestamp.

Client can compose words from tokens. For English, the first token of  new word starts with space, while for Chinese, each token is a word.

### CODE REVIEW

File: sherpa-ncnn/c-api/c-api.h [hunk 1]
```diff
@@ -114,7 +114,9 @@ typedef struct SherpaNcnnRecognizerConfig {
 
 typedef struct SherpaNcnnResult {
   const char *text;
-  // TODO: Add more fields
+  const char *tokens;
```
Comment (csukuangfj) [inline sherpa-ncnn/c-api/c-api.h:117]: Could you add some comments for tokens, timestamps, and count?

File: sherpa-ncnn/c-api/c-api.h [hunk 2]
```diff
@@ -114,7 +114,9 @@ typedef struct SherpaNcnnRecognizerConfig {
...
+  float* timestamps;
+  int count;
```
Comment (csukuangfj) [inline sherpa-ncnn/c-api/c-api.h:119]:
```suggestion
  int32_t count;
```

File: sherpa-ncnn/csrc/recognizer.h [hunk 3]
```diff
@@ -34,9 +34,10 @@
 namespace sherpa_ncnn {
 
 struct RecognitionResult {
-  std::vector<int32_t> tokens;
   std::string text;
   std::vector<float> timestamps;
+  std::vector<int32_t> tokens;
+  std::vector<std::string> stokens;
```
Comment (csukuangfj) [inline sherpa-ncnn/csrc/recognizer.h:40]: Could you document stokens?

File: sherpa-ncnn/csrc/recognizer.cc [hunk 4]
```diff
@@ -163,7 +176,10 @@ class Recognizer::Impl {
     DecoderResult decoder_result = s->GetResult();
     decoder_->StripLeadingBlanks(&decoder_result);
 
-    return Convert(decoder_result, sym_);
+    // Those 2 parameter figured out from sherpa source code
```
Comment (csukuangfj) [inline sherpa-ncnn/csrc/recognizer.cc:179]:
```suggestion
    // Those 2 parameters are figured out from sherpa source code
```

File: sherpa-ncnn/csrc/stream.cc [hunk 5]
```diff
@@ -50,7 +50,11 @@ class Stream::Impl {
 
   int32_t &GetNumProcessedFrames() { return num_processed_frames_; }
 
-  void SetResult(const DecoderResult &r) { result_ = r; }
+  void SetResult(const DecoderResult &r) { 
+	  int32_t offset = result_.frame_offset;
+	  result_ = r; 
+	  result_.frame_offset = offset;
```
Comment (csukuangfj) [inline sherpa-ncnn/csrc/stream.cc:56]: Can we use `r.offset` directly?

File: sherpa-ncnn/c-api/c-api.cc [hunk 6]
```diff
@@ -117,17 +117,39 @@ void Decode(SherpaNcnnRecognizer *p, SherpaNcnnStream *s) {
...
   r->text = new char[text.size() + 1];
   std::copy(text.begin(), text.end(), const_cast<char *>(r->text));
   const_cast<char *>(r->text)[text.size()] = 0;
+  r->count = res.tokens.size();
+  if (r->count > 0) {
+    // Each word ends with nullptr
+    r->tokens = new char[text.size() + r->count];
+    memset(reinterpret_cast<void*>(r->tokens), 0, text.size() + r->count);
+    r->timestamps = new float[r->count];
+    int pos = 0;
```
Comment (csukuangfj) [inline sherpa-ncnn/c-api/c-api.cc:132]:
```suggestion
    int32_t pos = 0;
```

File: sherpa-ncnn/c-api/c-api.cc [hunk 7]
```diff
@@ -117,17 +117,39 @@ void Decode(SherpaNcnnRecognizer *p, SherpaNcnnStream *s) {
...
+    for (int i = 0; i < r->count; ++i) {
```
Comment (csukuangfj) [inline sherpa-ncnn/c-api/c-api.cc:133]:
```suggestion
    for (int32_t i = 0; i < r->count; ++i) {
```

File: sherpa-ncnn/csrc/stream.cc (hunk 5 above)
Comment (leohuang2013) [reply sherpa-ncnn/csrc/stream.cc:56]: Not sure about this. If change to r.offset. Need test again.

File: sherpa-ncnn/c-api/c-api.h [hunk 8]
```diff
@@ -113,10 +113,19 @@ typedef struct SherpaNcnnRecognizerConfig {
...
+  // Recognized text
   const char *text;
+
+  // Pointer to continuous memory which holds string based tokens
+  // which are seperated by \0
   const char *tokens;
+
+  // Pointer to continuous memory which holds timestamps which
+  // are seperated by \0
   float* timestamps;
```
Comment (csukuangfj) [inline sherpa-ncnn/c-api/c-api.h:125]: Please fix the doc about `timestamps`. It is not a string and does not contain `\0`.

**Comment by csukuangfj** [comment]
Thank you!

**Pull request merged**